lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php

```diff
@@ -0,0 +1,168 @@
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\Diagnostic;
+use Phpactor\LanguageServerProtocol\DiagnosticSeverity;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServer\Core\CodeAction\CodeActionProvider;
+use Phpactor\LanguageServer\Core\Diagnostics\DiagnosticsProvider;
+use Phpactor\TextDocument\ByteOffset;
+use function Amp\call;
+use function array_filter;
+
+class GenerateMethodProvider implements DiagnosticsProvider, CodeActionProvider
+{
+    public const KIND = 'quickfix.generate_method';
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(Parser $parser)
+    {
+        $this->parser = $parser;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function kinds(): array
+    {
+        return [
+             self::KIND
+         ];
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function provideDiagnostics(TextDocumentItem $textDocument): Promise
+    {
+        return new Success($this->getDiagnostics($textDocument));
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function provideActionsFor(TextDocumentItem $textDocument, Range $range): Promise
+    {
+        return call(function () use ($textDocument, $range) {
+            if ($range->start != $range->end) {
+                return [];
+            }
+
+            $diagnostics = $this->getDiagnostics($textDocument);
+
+            if (empty($diagnostics)) {
+                return [];
+            }
+
+            $diagnostics = array_values(
+                array_filter($diagnostics, function (Diagnostic $diag) use ($range) {
+                    return (
+                        $diag->range->start->line <= $range->start->line &&
+                        $diag->range->start->character <= $range->start->character &&
+                        $diag->range->end->line >= $range->end->line &&
+                        $diag->range->end->character >= $range->end->character
+                    );
+                })
+            );
+
+            if (count($diagnostics) === 0) {
+                return [];
+            }
+
+            return [
+                CodeAction::fromArray([
+                    'title' =>  'Generate method (if not exists)',
+                    'kind' => self::KIND,
+                    'diagnostics' => $diagnostics,
+                    'command' => new Command(
+                        'Generate method',
+                        GenerateMethodCommand::NAME,
+                        [
+                            $textDocument->uri,
+                            PositionConverter::positionToByteOffset($range->start, $textDocument->text)->toInt()
+                        ]
+                    )
+                ])
+            ];
+        });
+    }
+    /**
+     * @return array<Diagnostic>
+     */
+    private function getDiagnostics(TextDocumentItem $textDocument): array
```

Comment:
Diagnostics are there, but they are not used in the LSP as they would underline every method call.

Reply:
Why not detect if the method exists or not?

Reply:
a) Was worried about performance
b) Wanted to get the other part running
c) Didn't know what was the best way to do that.

If you can help with c), that'll be great.